### PR TITLE
Native Windows urbit binaries using MingW

### PR DESCRIPTION
**Pull request by locpyl-tidnyd**
This PR enables C urbit binaries to build and run natively on Windows using MingW. 

Remaining work:

- [x] GitHub action to build the binaries
- [x] ~~graceful exit (|exit etc) does not work~~
- [x] ~~binaries are not completely standalone~~
- [x] ~~CLI terminal is not available~~
- [x] ~~daemon mode is disabled~~

To try it out, install the MSYS2 environment, check out the source or download a [tarball](https://github.com/locpyl-tidnyd/urbit/archive/refs/heads/mingw-port.zip), open a MingW-64 shell (not a MSYS shell, they are not interchangable), change to the `pkg/urbit` directory, and run first `./configure` to prepare the build environment and then `make build/urbit build/urbit-worker` to build. If you don't have the binary pill file checked out, `./configure` will detect that and download it. The resulting binaries are standalone and run as-is (without copying dlls) both from the MingW shell and from Windows.

`./configure` can optionally use Cachix to store and fetch pre-built dependencies in a manner similar to Nix. To enable fetching pre-built dependencies, set the `CACHIX_CACHE` environment variable to the name of the cache you want to use. The default cache used by Urbit is `ares`, but it is not write-accessible from external PRs. This PR uses uses the cache `locpyl-tidnyd-test1`. To store the dependencies you build, put your access token into the `CACHIX_AUTH_TOKEN` environment variable.
```
export CACHIX_CACHE=cache-name
export CACHIX_AUTH_TOKEN=eyJh... # if you have write access
./configure
```

### CODE REVIEW

**Comment by Fang-** [comment]
> albeit for now without CLI terminal functionality

Is this running up against Window's lack of ansi terminal emulation and/or general compatibility in that area? Curious to hear what you think it would take to support this.

In any case: very cool stuff, thank you!

**Comment by ohAitch** [comment]
I think in that case you could run it in PuTTY? (Or VSCode or for that matter Windows Terminal)

**Comment by locpyl-tidnyd** [comment]
In more recent Windows 10, the default console actually supports VT escape characters as an opt-in, but the good old Win32 console functions should be enough to implement the terminal. In fact, I suspect that one could even get by with just `\r` and '\b` and turning off echo. The real answer is that I did not consider the terminal a high priority item. Everyone, myself included, uses herb/http via lens port to talk to ships running in the background anyway. Also, the code in term.c is rather hairy and I've never used termios/termcaps before, so I concentrated on [getting ships off the ground](https://istheshipstillstuck.com/) first.

**Comment by locpyl-tidnyd** [comment]
In more recent Windows 10, the default console actually supports VT escape characters as an opt-in, but the good old Win32 console functions should be enough to implement the terminal. In fact, I suspect that one could even get by with just `\r` and `\b` and turning off echo. The real answer is that I did not consider the terminal a high priority item. Everyone, myself included, uses herb/http via lens port to talk to ships running in the background anyway. Also, the code in term.c is rather hairy and I've never used termios/termcaps before, so I concentrated on [getting ships off the ground](https://istheshipstillstuck.com/) first.

**Comment by Fang-** [comment]
> the code in term.c is rather hairy

#4463 intends to make it a bit less hairy (and also refactors the terminal protocol a little bit). You may or may not want to develop/test against that if you end up putting in a lot of term.c work. Also happy to hop on a call or email thread or w/e and walk you through the codebase, if that's useful to you!

If you want to fully support all functionality, you'll need to get fancier than just `\r` `\b`. You can imagine getting smart about down-converting output that's too complex for Win32 consoles to properly render, but at that point you have to ask yourself what you want the terminal for in the first place.

**Comment by locpyl-tidnyd** [comment]
@Fang- I looked at what you are doing in term.c in #4463, and it seems to be all above the plumbing layer I am concerned with: the layer that provides you with fds/`uv_stream_t*`s that support VT escapes, and switches them to/from raw mode on request. I'm going to try to migrate it to `uv_tty_t` and let libuv do most of the heavy lifting. `uv_tty_t` does VT emulation itself when it detects that Windows doesn't understand it. The only problem place in term.c for me is the use of `write(2)` in `_term_spin_step()` at a time when the terminal is (I think) in raw mode, as escapes and unicode characters written past libuv may not be interpreted. It _sort of_ works as-is on Windows 10, but rewriting it so that it uses `uv_write` would be cleaner.

**Comment by locpyl-tidnyd** [comment]
@Fang- Thank you. I have since gone over term.c a couple times trying to refactor plumbing, and while I still don't understand the dill/drum logic, luckily I find that I don't really have to. I looked at what you are doing in term.c in #4463, and it seems to be all above the bottommost plumbing layer I am concerned with: the layer that provides you with fds/`uv_stream_t*`s that support VT escapes, and switches them to/from raw mode on request. I'm going to try to migrate it to `uv_tty_t` and let libuv do most of the heavy lifting. `uv_tty_t` does VT emulation itself when it detects that Windows doesn't understand it. The only problem place in term.c for me is the use of `write(2)` in `_term_spin_step()` at a time when the terminal is (I think) in raw mode, as escapes and unicode characters written past libuv may not be interpreted. It _sort of_ works as-is on Windows 10, but rewriting it so that it uses `uv_write` would be cleaner.

**Comment by joemfb** [comment]
@locpyl-tidnyd, the terminal is (almost) always in raw mode. The spinner was written that way (stealing the fd, writing synchronously) because it used to run on its own thread, and I haven't seen a need to change it.

If `term.c` gives you too much trouble, it might be reasonable for the first pass to just disable it entirely (as `-t` does).

**Comment by locpyl-tidnyd** [comment]
I've added console support that works in native Windows console. Console does weird stuff if urbit is started from a console emulator such as ConEmu, Msys or Cygwin, because an emulated console is completely different from a native Windows console, but I try to detect this situation and warn the user that it's not supported.

**Comment by locpyl-tidnyd** [comment]
The latest changes fix most of the remaining terminal issues and enable basic daemon mode. The situation is as follows:
- `-d` now makes the urbit process detach from the native Windows console, if any, and redirect standard handles that used to be attached to the console to `/dev/null`. (If output was redirected to a file or pipe, it is not affected.) `-d` does nothing if urbit is run from a console emulator. I could make it "fork" away (Windows does not have `fork(2)` so it's a bit more complex), but I don't think it's worth bothering with. urbit.exe can be launched directly from Run or from a shortcut, or, to avoid having its console blink even briefly, with `start /b` or suitable task scheduler options. If launching it with `CreateProcess` then `CREATE_NO_WINDOW` and `DETACHED_PROCESS` flags should behave as expected.
- Live terminal works okay in native Windows console and almost okay in console emulators that use named pipes to communicate with their "terminal" processes (some startup messages may be slightly garbled, so I've left the "experimental support" message in).
- The new job I added to the `build.yml` workflow builds static Windows binaries. It does not strip them of debugging symbols, which would make sense for release, but releasing is above my (present) pay grade.

Given all this, I am opening this PR for review.

**Comment by locpyl-tidnyd** [comment]
The latest changes fix most of the remaining terminal issues and enable basic daemon mode. The situation is as follows:
- `-d` now makes the urbit process detach from the native Windows console, if any, and redirect standard handles that used to be attached to the console to `/dev/null`. (If output was redirected to a file or pipe, it is not affected.) `-d` does nothing if urbit is run from a console emulator. I could make it "fork" away (Windows does not have `fork(2)` so it's a bit more complex), but I don't think it's worth bothering with. urbit.exe can be launched directly from Run or from a shortcut, or, to avoid having its console blink even briefly, with `start /b` or suitable task scheduler options. If launching it with `CreateProcess` then `CREATE_NO_WINDOW` and `DETACHED_PROCESS` flags should behave as expected.
- Live terminal works okay in native Windows console. It works _almost_ okay in console emulators that use named pipes to communicate with their "terminal" processes: a few startup messages may be slightly garbled, so I've left the "experimental support" message in.
- The new job I added to the `build.yml` workflow builds static Windows binaries. It does not strip them of debugging symbols, which would make sense only for release.

Given all this, I am opening this PR for review.

**Comment by locpyl-tidnyd** [comment]
Changed `-d` to work in the same manner on MingW as it does on other systems: the initial process creates a child that does the actual work, waits for it to boot, and then exits. This has enabled me to add a smoke test to `build.yml` which boots a fake `~zod` and shuts it down with a lens request.

One limitation, due to how Windows console handles work internally, is that if `-d` is used from the native Windows console and stdout/stderr are not redirected, then no ship output will be visible. This is awkward to work around, and daemon mode is intended for starting background processes anyway, so I consider having no output in this case to be an acceptable compromise.

**Comment by locpyl-tidnyd** [comment]
Changed `-d` to work in the same manner on MingW as it does on other systems: the initial process creates a child that does the actual work, waits for it to boot, and then exits. This has enabled me to add a smoke test to `build.yml` which boots a fake `~zod` and shuts it down with a lens request.

One limitation, due to how Windows console handles work internally, is that if `-d` is used from the native Windows console and stdout/stderr are not redirected, then no ship output will be visible. This is awkward to work around, and daemon mode is intended for starting background processes anyway, so I consider having no output in this case to be an acceptable compromise. If you want to see ship output in the native console, use `-t`.

**Review by joemfb — CHANGES_REQUESTED** [review]
My apologies for the long delay. This is excellent work.

I've tested this branch pretty extensively now, and both the build scripts and resulting binary are working quite well.

I've reviewed the changes above, requesting clarification around some of the more complicated pieces, suggested some small tweaks, and noted potential future improvements. Some further notes:

### build:

- the "poor man's nix shell" extends `*-sources.json` with a `mingw` object to configure the build. this probably precludes future use of `nim` to manage those sources (i've been updating them manually anyway, /cc @brendanhay)

### patches:

- libh2o is patched to workaround windows opposite `iovec` member order with a new `STRLIT` macro, and to disable functionality around `AF_UNIX` sockets
- lmdb is patched to fix it's error-logging string allocation

### known issues:

- `mmap()` with `MAP_FIXED` very rarely runs into issues due to ASLR (see https://stackoverflow.com/questions/41079506/how-to-reserve-memory-on-windows-and-later-map-files-into-it)
- the terminal does not work well inside terminal emulators (works great in `cmd.exe`, haven't tested Windows Terminal)
- ctrl-z (ungraceful exit) is not implemented. I don't think this is particularly urgent.

### libsigsegv replacement:

- this PR includes a custom SEH handler to replace our use of libsigsegv, resulting in massive performance improvements

before:

```
> :time =|(i=@ud |-(?:(=(i 0xff.ffff) ~ $(i +(i)))))
>=
[%took ~s5..c229]
> :time =|(i=@ud |-(?:(=(i 0xff.ffff) ~ $(i +(i)))))
>=
[%took ~s2..d5c8]

> :time (gulf 1 0xf.ffff)
>=
[%took ~s5..0cac]
> :time (gulf 1 0xf.ffff)
>=
[%took ~s0..8a08]

> :time (jam (gulf 1 0xf.ffff))
>=
[%took ~s29..9843]
> :time (jam (gulf 1 0xf.ffff))
>=
[%took ~s2..d245]

> :time (cue (jam (gulf 1 0xf.ffff)))
>=
[%took ~s38..d209]
> :time (cue (jam (gulf 1 0xf.ffff)))
>=
[%took ~s5..7cdf]
```

after:

```
> :time =|(i=@ud |-(?:(=(i 0xff.ffff) ~ $(i +(i)))))
>=
[%took ~s3..0489]
> :time =|(i=@ud |-(?:(=(i 0xff.ffff) ~ $(i +(i)))))
>=
[%took ~s2..fb4a]

> :time (gulf 1 0xf.ffff)
>=
[%took ~s0..af85]
> :time (gulf 1 0xf.ffff)
>=
[%took ~s0..9e9e]

> :time (jam (gulf 1 0xf.ffff))
>=
[%took ~s2..fbbc]
> :time (jam (gulf 1 0xf.ffff))
>=
[%took ~s2..d823]

> :time (cue (jam (gulf 1 0xf.ffff)))
>=
[%took ~s5..c382]
> :time (cue (jam (gulf 1 0xf.ffff)))
>=
[%took ~s5..a927]
```

(each expression was timed twice in a new process, the first time includes the cost of page-faults. note also that `:time` is a low-resolution benchmark of wall-clock time ...)

--------

Barring the minor improvements requested above, I think this PR is ready to be merged. The biggest obstacle to integration will be the coming terminal rewrite (#4463). I tried a merge locally, and the conflicts look quite manageable. But it's possible that there will be issues with new functionality (2-dimensional output / screen updates, log/slog handling). /cc @Fang-

File: nix/sources-mingw.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    "ge-additions": {
+        "mingw": {
+            "make": "CFLAGS=-I../ed25519"
+        }
+    },
+    "libaes_siv": {
+        "mingw": {
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply nix/sources-mingw.json]: Fixed.

File: pkg/urbit/compat/mingw/compat.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+int kill(pid_t pid, int signum);
+
+#define SIGUSR1   10
+#define SIGALRM   14
+#define SIGVTALRM 26
+#define SIGSTK    31
+#define SIG_COUNT 32
+
+#endif//_MINGW_IO_H
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/compat/mingw/compat.h]: Fixed.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -116,3 +116,24 @@ jobs:
...
+
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          lfs: true
+
+      # echo suppresses pacman prompt
+      - run: echo|./configure
+        env:
+          CACHIX_CACHE: locpyl-tidnyd-test1
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply .github/workflows/build.yml:135]: Please do this when merging, as I am using this temporary cache while I'm still working on the PR.

File: pkg/ent/configure [hunk 4]
```diff
@@ -7,6 +7,7 @@ log () {
 for impl in ENT_GETENTROPY_UNISTD \
             ENT_GETENTROPY_SYSRANDOM \
             ENT_GETRANDOM_SYSCALL \
+            ENT_GETENTROPY_BCRYPTGENRANDOM \
             ENT_DEV_URANDOM
 do
     export IMPL=$impl
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/ent/configure:13]: Added `.gitignore` lines in 839a68e.

File: pkg/urbit/compat/mingw/argon2u.patch [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+     // because this generates WASM error:
+@@ -416,7 +416,7 @@ int encode_string(char *dst, size_t dst_len, argon2_context *ctx,
+ #define SX(x)                                                                  \
+     do {                                                                       \
+         char tmp[30];                                                          \
+-        itoa(x, tmp);                                                          \
++        encode_decimal(x, tmp);                                                \
+         SS(tmp);                                                               \
+     } while ((void)0, 0)
+
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/compat/mingw/argon2u.patch]: 3559394 updates the argon2u dependency and removes the patch.

File: pkg/urbit/compat/mingw/seh_handler_decorator.mk [hunk 6]
```diff
@@ -0,0 +1,7 @@
+sehdexe := build/seh_handler_decorator
+CCDEPS  := $(CCDEPS) $(sehdexe)
+CCEXTRA  = -E -o -|$(sehdexe) $<|$(CC) $(CFLAGS) -x cpp-output -
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/compat/mingw/seh_handler_decorator.mk:20]: 4ecfb14 comments what is happening. `-x cpp-output` tells `$(CC)` to treat its standard input as preprocessor output, as it cannot deduce this from the input filename.

File: pkg/urbit/compat/mingw/seh_handler_decorator.cc [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            fputc(line[i], stdout);
+            continue;
+        emit_handler:
+            fputs("{__asm__(\".seh_handler u3_exception_handler,@except\\n\");", stdout);
+            state = INIT;
+        }
+    }
+
+    return 0;
+}
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/compat/mingw/seh_handler_decorator.cc:65]: Yes. I have added [some comments](https://github.com/urbit/urbit/blob/4ecfb14cbfbbc6cd6d0ac2180298cb2ce5a82ff4/pkg/urbit/compat/mingw/seh_handler_decorator.mk) about both the build process and the technique in commit 4ecfb14:
```
# This include file injects a step that transforms vere C source to register
# a SEH exception handler for each function that is declared in a .c file.
# It inserts a .seh_handler directive into each function body with __asm__.
# This directive affects the x64 unwind tables (.pdata and .xdata sections)
# emitted by the compiler.
#
# See gas/config/obj-coff-seh.h in binutils source for .seh_handler, and
# https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64
# for a description of how stack unwinding and SEH work in x64 Windows.
```

**Comment by locpyl-tidnyd** [comment]
Ctrl-Break is the "ungraceful exit" key in Windows console. Windows uses `^Z` as the end-of-file character where Unix uses `^D`.

File: pkg/urbit/compat/mingw/poor-mans-nix-shell.sh [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+buildnixdep () {
+  echo Building dependency $key...
+  local cache=https://app.cachix.org/api/v1/cache/${CACHIX_CACHE-}
+  local hash=
+  if [ -n "$url" ]
+  then
+    hash=${hdeps[$key]}
+    dir=../$hash-$key
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/compat/mingw/poor-mans-nix-shell.sh]: f869bd1 does this.

File: pkg/urbit/Makefile [hunk 9]
```diff
@@ -74,9 +77,9 @@ build/urbit-worker: $(common_objs) $(worker_objs)
 	@mkdir -p ./build
 	@$(CC) $^ $(LDFLAGS) -o $@
 
-%.o: %.c $(headers)
+%.o: %.c $(headers) $(CCDEPS)
 	@echo CC $<
-	@$(CC) -I./include $(CFLAGS) -c $< -o $@
+	@$(CC) -I./include $(CFLAGS) $< $(CCEXTRA) -c -o $@
```
[the remark this answers is not in the document]
Comment (locpyl-tidnyd) [reply pkg/urbit/Makefile:84]: Added a comment in 4ecfb14.

**Review by joemfb — APPROVED** [review]
LGTM

**Review by locpyl-tidnyd — COMMENTED** [review]
Commit message fail: should have been "put downloaded dependencies under $NIX_STORE (defaults to ...)"

**Pull request merged**